### PR TITLE
 Create TriangleTestGenerator. Update TriangleTest. Refs #370, #331

### PR DESCRIPTION
**Pull request by ricemery**
*  Create TriangleTestGenerator. 
* Update TriangleTest
* Rework example.scala to conform to new class/function signature. The old example used enumeration to represent the triangle type. A triangle could only be a single type. Only being a single type is invalid according to test - "equilateral triangles are also isosceles".  I removed the enumeration from the example. Now the class supports `isosceles`, `equilateral` and `scalene` functions as spec'd in the canonical-data.json.

### CODE REVIEW

File: testgen/src/main/scala/TriangleTestGenerator.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      arg match {
+        case vals: List[_] => vals.mkString(", ")
+        case _ => throw new IllegalStateException
+      }
+    }
+
+    def sutArgs(parseResult: CanonicalDataParser.ParseResult, argNames: String*): String =
+          argNames map (name => mapArgToString(parseResult(name))) mkString
+
+    def fromLabeledTest(argNames: String*): ToTestCaseData =
```
Comment (abo64) [inline testgen/src/main/scala/TriangleTestGenerator.scala]: Have you tried `fromLabeledTestAlt` as in [BeerSongTestGenerator](https://github.com/exercism/scala/blob/master/testgen/src/main/scala/BeerSongTestGenerator.scala#L10)? Might simplify matters a bit?

Comment (abo64) [inline testgen/src/main/scala/TriangleTestGenerator.scala]: I just did a quick test with
```scala
val code = TestSuiteBuilder.build(file,
  fromLabeledTestAlt("equilateral" -> Seq("sides"), "isosceles" -> Seq("sides"),  "scalene" -> Seq("sides")))
```
and it seems to work well.

Comment (abo64) [inline testgen/src/main/scala/TriangleTestGenerator.scala]: But then the output assumes a stateless object - which might be a considerable alternative? For example:
```scala
test("true if all sides are equal") { 
    Triangle.equilateral(List(2, 2, 2)) should be (true)
  }
```

Comment (ErikSchierboom) [inline testgen/src/main/scala/TriangleTestGenerator.scala]: I'm fine with using a stateless object here, it's not that it requires a lot of fancy logic for which an object could be useful.

Comment (ricemery) [inline testgen/src/main/scala/TriangleTestGenerator.scala]: I don't have a real strong opinion. But, I would be more comfortable with a stateless object if there was only a single method returning a type. But, in this case there are multiple methods potentially operating on the same values. And, a triangle can be multiple types..

At the end of the day - what is probably most important is what is the goal of the exercise? What is the exercise meaning to teach/re-enforce?  
It is always a bit difficult to figure out what a generic exercise is attempting to teach.

Comment (abo64) [inline testgen/src/main/scala/TriangleTestGenerator.scala]: Your argument about the multiple methods makes sense to me.
So one would have to modify the generator code in in a similar way as you did.

But then we might as well just take your code. :)

Comment (ricemery) [inline testgen/src/main/scala/TriangleTestGenerator.scala]: I think for now, I am going to leave the test generator and the definition as is.

I like your more simplified generator.

```
val code = TestSuiteBuilder.build(file,
  fromLabeledTestAlt("equilateral" -> Seq("sides"), "isosceles" -> Seq("sides"),  "scalene" -> Seq("sides")))
```
But... The generator I wrote adds additional info to the test descriptions that I thinks is handy.
For example. 

```
  false if no sides are equal
``` 

becomes 

```
  isosceles - false if no sides are equal
```
when my generator runs. I don't think the original description is all that helpful. This is probably better to submit new descriptions to the canonical json. But, for now having the generator add the extra info to the beginning of the comments is useful IMHO.

**Comment by abo64** [comment]
thanks @ricemery

**Pull request merged**